File: client/country_select.py
"""Entry-country selector helpers (flags, default Iceland/IS, Connect gate).

Used by the main Connect shell dropdown (not Settings-only). Pure and
unit-testable without GUI.

Honesty: emoji flags may not render on every Windows font — option labels
always include country code + name; flag glyph is best-effort decoration.
Prefer Segoe UI Emoji on Windows so regional-indicator flags show in the menu.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Sequence

from .multihop import (
    COUNTRY_DE,
    COUNTRY_IS,
    COUNTRY_RO,
    COUNTRY_US,
    DEFAULT_ENTRY_COUNTRY,
    PRODUCT_COUNTRY_CATALOG,
    CountryNode,
    normalize_entry_country,
    product_country_catalog,
)

# Regional-indicator flag sequences (IS / DE live; US retired). Safe as unicode text.
_FLAG_BY_CODE: dict[str, str] = {
    COUNTRY_IS: "\U0001f1ee\U0001f1f8",  # 🇮🇸
    COUNTRY_DE: "\U0001f1e9\U0001f1ea",  # 🇩🇪
    COUNTRY_US: "\U0001f1fa\U0001f1f8",  # 🇺🇸
    # Retired RO flag kept for any legacy admin display only
    COUNTRY_RO: "\U0001f1f7\U0001f1f4",  # 🇷🇴
}


@dataclass(frozen=True)
class CountryOption:
    """One dropdown row for the main-shell entry-country control."""

    code: str
    name: str
    flag: str
    host: str = ""

    def label(self) -> str:
        """Display string: flag + name + code (flag may be empty on bare fonts)."""
        flag = (self.flag or "").strip()
        name = (self.name or self.code or "").strip()
        code = (self.code or "").strip().upper()
        if flag:
            return f"{flag}  {name} ({code})"
        return f"{name} ({code})"

    def to_dict(self, *, admin: bool = False) -> dict[str, str]:
        """UI/JSON dict. Non-admin omits residual monopin *host* (presentation hygiene)."""
        from client.residual_public import public_country_option_dict

        return public_country_option_dict(
            code=self.code,
            name=self.name,
            flag=self.flag,
            host=self.host,
            admin=admin,
        )


def country_flag_emoji(code: str | None) -> str:
    """Flag glyph for a catalog country code (empty if unknown)."""
    c = (code or "").strip().upper()
    if c in _FLAG_BY_CODE:
        return _FLAG_BY_CODE[c]
    # Best-effort regional indicators for 2-letter codes
    if len(c) == 2 and c.isalpha():
        try:
            return "".join(chr(0x1F1E6 + (ord(ch) - ord("A"))) for ch in c)
        except Exception:  # noqa: BLE001
            return ""
    return ""


def parse_catalog_country_code(
    raw: str | None,
    *,
    catalog: Sequence[CountryNode] | None = None,
) -> str | None:
    """Return catalog code if *raw* is a known entry country; else None.

    Unlike :func:`normalize_entry_country`, does **not** default unknown/empty
    to DE — used for Connect gate and strict validation.
    """
    raw_s = (raw or "").strip()
    if not raw_s:
        return None
    # Reuse alias table via normalize, then verify membership
    code = normalize_entry_country(raw_s)
    # If input was garbage, normalize returns DEFAULT — only accept if raw
    # maps intentionally (alias or exact catalog code).
    cat = list(catalog) if catalog is not None else list(product_country_catalog())
    codes = {str(getattr(n, "code", "") or "").strip().upper() for n in cat}
    upper = raw_s.upper()
    aliases = {
        "ICELAND": COUNTRY_IS,
        "IS": COUNTRY_IS,
        "GERMANY": COUNTRY_DE,
        "DE": COUNTRY_DE,
        "DEU": COUNTRY_DE,
        "DEUTSCHLAND": COUNTRY_DE,
        # Stale US/RO are not catalog members (normalize maps → DE)
        "UNITED STATES": COUNTRY_US,
        "UNITED STATES OF AMERICA": COUNTRY_US,
        "USA": COUNTRY_US,
        "US": COUNTRY_US,
        "AMERICA": COUNTRY_US,
        "ROMANIA": COUNTRY_RO,
        "RO": COUNTRY_RO,
        "ROU": COUNTRY_RO,
    }
    want = aliases.get(upper, upper)
    if want in codes:
        return want
    # normalize may have mapped — only accept if want was a known alias path
    if code in codes and upper in aliases:
        return code
    return None


def default_entry_country() -> str:
    """Product default entry: Germany (DE) on every client."""
    return DEFAULT_ENTRY_COUNTRY


def default_entry_reason() -> str:
    """Reason token when empty selection falls back to product default."""
    if DEFAULT_ENTRY_COUNTRY == COUNTRY_DE:
        return "default_germany"
    if DEFAULT_ENTRY_COUNTRY == COUNTRY_US:
        return "default_united_states"
    if DEFAULT_ENTRY_COUNTRY == COUNTRY_IS:
        return "default_iceland"
    return "default_entry"


def resolve_entry_country_selection(
    raw: str | None,
    *,
    catalog: Sequence[CountryNode] | None = None,
    allow_default: bool = True,
) -> tuple[bool, str, str]:
    """Resolve selection for Connect.

    Returns ``(ok, code, reason)``:
    - empty/None + allow_default → (True, IS, \"default_iceland\")
    - valid catalog → (True, code, \"ok\")
    - empty without default / unknown → (False, \"\", reason)
    """
    raw_s = (raw or "").strip()
    if not raw_s:
        if allow_default:
            return True, default_entry_country(), default_entry_reason()
        return False, "", "missing_entry_country"
    code = parse_catalog_country_code(raw_s, catalog=catalog)
    if code is None:
        return False, "", "invalid_entry_country"
    return True, code, "ok"


def entry_country_allows_connect(
    raw: str | None,
    *,
    catalog: Sequence[CountryNode] | None = None,
    allow_default: bool = True,
) -> bool:
    """True when Connect may dial for this selection (valid or product default)."""
    ok, _code, _reason = resolve_entry_country_selection(
        raw, catalog=catalog, allow_default=allow_default
    )
    return ok


def catalog_country_options(
    catalog: Sequence[CountryNode] | None = None,
) -> list[CountryOption]:
    """Dropdown options from product residual catalog (flags + labels)."""
    cat = list(catalog) if catalog is not None else list(product_country_catalog())
    out: list[CountryOption] = []
    seen: set[str] = set()
    for n in cat:
        code = str(getattr(n, "code", "") or "").strip().upper()
        if not code or code in seen:
            continue
        seen.add(code)
        name = str(getattr(n, "name", "") or code).strip() or code
        host = str(getattr(n, "host", "") or "").strip()
        flag = country_flag_emoji(code)
        # Always expose a non-empty flag field for catalog codes (emoji or fallback)
        if not flag:
            flag = f"[{code}]"
        out.append(
            CountryOption(
                code=code,
                name=name,
                flag=flag,
                host=host,
            )
        )
    if not out:
        # Fail soft: still offer product default IS
        code = default_entry_country()
        out.append(
            CountryOption(
                code=code,
                name="Iceland" if code == COUNTRY_IS else code,
                flag=country_flag_emoji(code) or f"[{code}]",
                host="",
            )
        )
    return out


def option_label_for_code(
    code: str | None,
    *,
    catalog: Sequence[CountryNode] | None = None,
) -> str:
    """Human label for a stored code (product default IS when empty)."""
    ok, resolved, _ = resolve_entry_country_selection(
        code, catalog=catalog, allow_default=True
    )
    use = resolved if ok else default_entry_country()
    for opt in catalog_country_options(catalog):
        if opt.code == use:
            return opt.label()
    return f"{country_flag_emoji(use)}  {use}".strip()


def label_to_country_code(
    label: str | None,
    *,
    catalog: Sequence[CountryNode] | None = None,
) -> str | None:
    """Map a dropdown label back to catalog code (or parse code from text)."""
    text = (label or "").strip()
    if not text:
        return None
    for opt in catalog_country_options(catalog):
        if text == opt.label() or text == opt.code or text.endswith(f"({opt.code})"):
            return opt.code
    # Last resort: strict parse
    return parse_catalog_country_code(text, catalog=catalog)

File: client_app/lib/country_select.dart
/// Entry-country selector helpers (flags, Germany/DE default, Connect gate).
///
/// Catalog mirrors [client/multihop.py] PRODUCT_COUNTRY_CATALOG (IS/DE only).
library;

/// Product default residual entry (empty prefs / fresh install).
const String kDefaultEntryCountry = 'DE';
const String kCountryIceland = 'IS';
const String kCountryGermany = 'DE';
const String kCountryUnitedStates = 'US'; // retired — normalize maps US → DE
/// Retired residual peer codes — [normalizeEntryCountry] maps RO/US → default DE.
const String kCountryRomania = 'RO';

/// Product residual entry peers (code, name, flag emoji, monopin host).
class CountryOption {
  final String code;
  final String name;
  final String flag;
  final String host;

  const CountryOption({
    required this.code,
    required this.name,
    required this.flag,
    this.host = '',
  });

  /// User-facing label never includes residual monopin IP (dial host is private).
  String get label => flag.isEmpty ? '$name ($code)' : '$flag  $name ($code)';
}

/// Shipped catalog — keep aligned with client.multihop PRODUCT_COUNTRY_CATALOG.
///
/// [host] is the private dial address used by residual Connect/HELLO only.
/// UI must use [label] / [code] / [name] — never paint [host] in user surfaces.
const List<CountryOption> kProductCountryCatalog = [
  CountryOption(
    code: kCountryIceland,
    name: 'Iceland',
    flag: '🇮🇸',
    host: '82.221.101.241',
  ),
  CountryOption(
    code: kCountryGermany,
    name: 'Germany',
    flag: '🇩🇪',
    host: '178.105.187.178',
  ),
];

String defaultEntryCountry() => kDefaultEntryCountry;

/// Strict parse — unknown/empty/stale RO → null (does not default).
String? parseCatalogCountryCode(String? raw) {
  final upper = (raw ?? '').trim().toUpperCase();
  if (upper.isEmpty) return null;
  // Stale RO/US prefs are not catalog members (normalize maps to DE)
  if (upper == 'RO' ||
      upper == 'ROMANIA' ||
      upper == 'ROU' ||
      upper == 'US' ||
      upper == 'USA' ||
      upper == 'UNITED STATES' ||
      upper == 'UNITED STATES OF AMERICA' ||
      upper == 'AMERICA') {
    return null;
  }
  const aliases = {
    'ICELAND': kCountryIceland,
    'IS': kCountryIceland,
    'GERMANY': kCountryGermany,
    'DE': kCountryGermany,
    'DEU': kCountryGermany,
    'DEUTSCHLAND': kCountryGermany,
  };
  final want = aliases[upper] ?? upper;
  for (final o in kProductCountryCatalog) {
    if (o.code == want) return o.code;
  }
  return null;
}

/// Normalize: empty/unknown/stale RO/US → Germany/DE (durable prefs default).
String normalizeEntryCountry(String? raw) {
  final c = parseCatalogCountryCode(raw);
  return c ?? kDefaultEntryCountry;
}

/// Resolve for Connect: (ok, code, reason).
({bool ok, String code, String reason}) resolveEntryCountrySelection(
  String? raw, {
  bool allowDefault = true,
}) {
  final trimmed = (raw ?? '').trim();
  if (trimmed.isEmpty) {
    if (allowDefault) {
      return (
        ok: true,
        code: kDefaultEntryCountry,
        reason: 'default_germany',
      );
    }
    return (ok: false, code: '', reason: 'missing_entry_country');
  }
  final code = parseCatalogCountryCode(trimmed);
  if (code == null) {
    // Stale RO / unknown: when allowDefault, map to product default DE
    if (allowDefault) {
      final upper = trimmed.toUpperCase();
      if (upper == 'RO' || upper == 'ROMANIA' || upper == 'ROU') {
        return (
          ok: true,
          code: kDefaultEntryCountry,
          reason: 'stale_ro_to_default_germany',
        );
      }
      if (upper == 'US' ||
          upper == 'USA' ||
          upper == 'UNITED STATES' ||
          upper == 'UNITED STATES OF AMERICA' ||
          upper == 'AMERICA') {
        return (
          ok: true,
          code: kDefaultEntryCountry,
          reason: 'stale_us_to_default_germany',
        );
      }
      // Other unknown → product default
      return (
        ok: true,
        code: kDefaultEntryCountry,
        reason: 'unknown_to_default_germany',
      );
    }
    return (ok: false, code: '', reason: 'unknown_entry_country');
  }
  return (ok: true, code: code, reason: 'catalog');
}

CountryOption? countryOptionForCode(String? code) {
  final c = normalizeEntryCountry(code);
  for (final o in kProductCountryCatalog) {
    if (o.code == c) return o;
  }
  return null;
}

/// Product exit hop for multi-hop residual (Germany monopin).
const String kProductExitHost = '178.105.187.178';

/// Residual dial host for entry *code* (+ multi-hop when on).
///
/// Multi-hop ON: dial [kProductExitHost] (DE) when entry is not already DE;
/// if entry is DE, stay on DE (exit == entry).
String residualHostForEntryCountry(
  String? code, {
  bool multiHop = false,
}) {
  final c = normalizeEntryCountry(code);
  if (multiHop) {
    // Residual-via-exit: dial DE exit when entry is not DE.
    if (c != kCountryGermany) {
      return kProductExitHost;
    }
  }
  for (final o in kProductCountryCatalog) {
    if (o.code == c) return o.host;
  }
  return kProductExitHost; // default DE
}

/// Other catalog residual monopin hosts for wipe-drain Connect failover.
///
/// Never includes [excluding] (preferred entry). Order follows product catalog.
List<String> alternateResidualHosts({String excluding = ''}) {
  final skip = excluding.trim();
  return [
    for (final o in kProductCountryCatalog)
      if (o.host.isNotEmpty && o.host != skip) o.host,
  ];
}

/// Public ElGamal pin basename for residual HELLO to *host*.
String residualNodePubNameForHost(String host) {
  final h = host.trim();
  for (final o in kProductCountryCatalog) {
    if (o.host == h) {
      switch (o.code) {
        case kCountryIceland:
          return 'node_elgamal.pub';
        case kCountryGermany:
          return 'de_node_elgamal.pub';
      }
    }
  }
  if (h == kProductExitHost || h == '178.105.187.178') {
    return 'de_node_elgamal.pub';
  }
  // Retired US monopin host — heal to DE pin
  if (h == '5.161.242.85') return 'de_node_elgamal.pub';
  if (h == '82.221.101.241') return 'node_elgamal.pub';
  // Stale RO host still maps to exit pin file (now DE content) for heal path only.
  if (h == '185.146.232.107') return 'exit_node_elgamal.pub';
  return 'node_elgamal.pub';
}

bool entryCountryAllowsConnect(
  String? raw, {
  bool allowDefault = true,
}) {
  return resolveEntryCountrySelection(raw, allowDefault: allowDefault).ok;
}

/// User-facing residual entry label — never a monopin IPv4.
String residualEntryPublicLabel([String? code]) {
  final o = countryOptionForCode(code);
  final name = (o?.name ?? '').trim();
  final c = (o?.code ?? kDefaultEntryCountry).trim();
  if (name.isEmpty) return c;
  return '$name ($c)';
}

/// Map dial host monopin → public country label (for status UI redaction).
String publicLabelForResidualHost(String? host) {
  final h = (host ?? '').trim();
  if (h.isEmpty) return 'VPN node';
  for (final o in kProductCountryCatalog) {
    if (o.host.isNotEmpty && (h == o.host || h.endsWith(o.host))) {
      return residualEntryPublicLabel(o.code);
    }
  }
  if (RegExp(r'^\d+\.\d+\.\d+\.\d+$').hasMatch(h)) {
    return 'VPN node';
  }
  return 'VPN node';
}
